internal/datanode/buffer.go

```diff
@@ -17,13 +17,220 @@
 package datanode
 
 import (
+	"container/heap"
 	"errors"
+	"fmt"
+	"github.com/milvus-io/milvus-proto/go-api/schemapb"
+	"github.com/milvus-io/milvus/internal/metrics"
 	"math"
+	"sync"
 
 	"github.com/milvus-io/milvus/internal/proto/datapb"
 	"github.com/milvus-io/milvus/internal/storage"
 )
 
+// DelBufferManager is in charge of managing insertBuf and delBuf from an overall prospect
+// not only controlling buffered data size based on every segment size, but also triggering
+// insert/delete flush when the memory usage of the whole manager reach a certain level.
+// but at the first stage, this struct is only used for delete buff
+type DelBufferManager struct {
+	delBufMap     sync.Map // map[segmentID]*DelDataBuf
+	delMemorySize int64
+	delBufHeap    *PriorityQueue
+}
+
+func (bm *DelBufferManager) SegDelBufMemSize(segID UniqueID) int64 {
+	if delDataBuf, ok := bm.delBufMap.Load(segID); ok {
+		return delDataBuf.(*DelDataBuf).item.memorySize
+	}
+	return 0
+}
+
+func (bm *DelBufferManager) GetEntriesNum(segID UniqueID) int64 {
+	if delDataBuf, ok := bm.delBufMap.Load(segID); ok {
+		return delDataBuf.(*DelDataBuf).GetEntriesNum()
+	}
+	return 0
+}
+
+// Store :the method only for unit test
+func (bm *DelBufferManager) Store(segID UniqueID, delDataBuf *DelDataBuf) {
+	bm.delBufMap.Store(segID, delDataBuf)
+}
+
+func (bm *DelBufferManager) StoreNewDeletes(segID UniqueID, pks []primaryKey,
+	tss []Timestamp, tr TimeRange) {
+	//1. load or create delDataBuf
+	var delDataBuf *DelDataBuf
```

Comment:
all of the functions are not protected by a lock, is that right?

Reply:
According to Congqi, all operations related to del/insert buf will be executed inside one channel and one groutine, so they will be run in serial sequence